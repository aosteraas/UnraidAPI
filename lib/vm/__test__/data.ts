export const vmHtml = `<tr parent-id="0" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-1d608caf-a589-395f-121a-c6c70f9abac6" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-0")'
          title="click for more VM info"
          >Lubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-1d608caf-a589-395f-121a-c6c70f9abac6" style="cursor: pointer"
      >2</a
    >
  </td>
  <td>4096M</td>
  <td title="Current physical size: 20G">1 / 20G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Lubuntu"
      title="Toggle VM autostart"
      uuid="1d608caf-a589-395f-121a-c6c70f9abac6"
    />
  </td>
</tr>
<tr child-id="0" id="name-0" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Lubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="1d608caf-a589-395f-121a-c6c70f9abac6"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Lubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="20G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">20G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="20G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>19G</td>
        </tr>
        <tr>
          <td>/mnt/user/isos/lubuntu-20.10-desktop-amd64.iso</td>
          <td>SATA</td>
          <td>2G</td>
          <td>2G</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
<tr parent-id="1" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-3c15a77b-7917-514e-91e5-3aa32c29f919" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-1")'
          title="click for more VM info"
          >Ubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-3c15a77b-7917-514e-91e5-3aa32c29f919" style="cursor: pointer"
      >1</a
    >
  </td>
  <td>1024M</td>
  <td title="Current physical size: 2G">1 / 2G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Ubuntu"
      title="Toggle VM autostart"
      uuid="3c15a77b-7917-514e-91e5-3aa32c29f919"
    />
  </td>
</tr>
<tr child-id="1" id="name-1" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Ubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="3c15a77b-7917-514e-91e5-3aa32c29f919"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Ubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="2G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">2G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="2G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>1M</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
addVMContext('Lubuntu','1d608caf-a589-395f-121a-c6c70f9abac6','Ubuntu','shutoff','','');addVMContext('Ubuntu','3c15a77b-7917-514e-91e5-3aa32c29f919','Ubuntu','shutoff','','');var
kvm=[];kvm.push({id:'1d608caf-a589-395f-121a-c6c70f9abac6',state:'shutoff'});kvm.push({id:'3c15a77b-7917-514e-91e5-3aa32c29f919',state:'shutoff'});`;
