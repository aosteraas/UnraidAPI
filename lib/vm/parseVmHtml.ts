import { parse } from 'node-html-parser';

interface ParsedVms {
  id: string;
  name: string;
  status: string;
  imageUrl: string;
  cores: string;
  memory: string;
  hddTotal: string;
}

export function parseVmHtml(rawHtml: string): {
  vms: ParsedVms[];
  extras: string;
} {
  const [core, extras] = rawHtml.split('\u0000');

  const document = parse(core);
  const parents = document.querySelectorAll('[parent-id]');
  const children = document.querySelectorAll('[child-id]');

  const vms = parents
    .map((parent) => {
      const id = parent.getAttribute('parent-id');
      const child = children.find((c) => c.getAttribute('child-id') === id);
      return { parent, child };
    })
    .map(({ parent, child }) => {
      const outer = parent.querySelector('.outer');
      const imageUrl = outer.querySelector('img.img').getAttribute('src');
      const id = outer
        .querySelector('.hand')
        .getAttribute('id')
        .replace('vm-', '');

      const inner = parent.querySelector('.inner');
      const name = inner.querySelector('a').innerText;
      const status = inner.querySelector('.state').innerText;
      const cores = parent.querySelector('[class^="vcpu"]*').innerText;
      const memory = parent.childNodes[3].innerText;
      const hddTotal = parent.querySelector('td[title]').innerText;

      return { id, name, status, imageUrl, cores, memory, hddTotal };
    });

  return { vms, extras: extras ?? '' };
}
