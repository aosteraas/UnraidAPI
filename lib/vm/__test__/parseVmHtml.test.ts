import { parseVmHtml } from '../parseVmHtml';
import { vmHtml } from './data';

test('Parses raw VM HTML into array of VM data and extras string', () => {
  const { vms, extras } = parseVmHtml(vmHtml);
  expect(vms).toMatchSnapshot();
  expect(extras).toMatchSnapshot();
});
