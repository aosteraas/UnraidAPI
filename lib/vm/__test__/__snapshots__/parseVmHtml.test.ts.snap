// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses raw VM HTML into array of VM data and extras string 1`] = `
Array [
  Object {
    "cores": "2",
    "hddTotal": "1 / 20G",
    "id": "1d608caf-a589-395f-121a-c6c70f9abac6",
    "imageUrl": "/plugins/dynamix.vm.manager/templates/images/ubuntu.png",
    "memory": "",
    "name": "Lubuntu",
    "status": "stopped",
  },
  Object {
    "cores": "1",
    "hddTotal": "1 / 2G",
    "id": "3c15a77b-7917-514e-91e5-3aa32c29f919",
    "imageUrl": "/plugins/dynamix.vm.manager/templates/images/ubuntu.png",
    "memory": "",
    "name": "Ubuntu",
    "status": "stopped",
  },
]
`;

exports[`Parses raw VM HTML into array of VM data and extras string 2`] = `""`;
